src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -60,80 +94,457 @@ W3DDisplay::W3DDisplay() :
-    // s_2DScene = nullptr;
-    // s_3DInterfaceScene->Release_Ref();
-    // s_3DInterfaceScene = nullptr;
-
-    // for (auto &light : m_myLight) {
-    //    light->Release_Ref();
-    //    light = nullptr;
-    //}
+    for (int i = 0; i < 16; i++) {
+        g_theDisplayStringManager->Free_Display_String(m_displayStrings[i]);
+    }
 
-    // PredictiveLODOptimizerClass::Free();
-    //// unk string 0xA48068 StringClass::Release_Resources();
-    // W3DShaderManager::Shutdown();
-    // s_assetManager->Free_Assets();
-    // delete s_assetManager;
-    //// WW3D::Shutdown();
-    //// WWMath::Shutdown();
-    //// Browser Shutdown();
-    //// delete g_theW3DFileSystem;
-
-    // TODO: Requires WW3D, WWMath, g_theW3DFileSystem, 0xA48068, Browser
-#ifdef GAME_DLL
-    Call_Method<void, W3DDisplay>(0x0073C470, this);
-#endif
+    delete m_2DRender;
+    m_2DRender = nullptr;
+
+    Delete_Views();
+
+    s_3DScene->Release_Ref();
+    s_3DScene = nullptr;
+
+    s_2DScene->Release_Ref();
+    s_2DScene = nullptr;
+
+    s_3DInterfaceScene->Release_Ref();
+    s_3DInterfaceScene = nullptr;
+
+    for (int i = 0; i < 4; i++) {
+        Ref_Ptr_Release(m_myLight[i]);
+    }
+
+    PredictiveLODOptimizerClass::Free();
+    // Debug_Statistics::Shutdown_Statistics
+    W3DShaderManager::Shutdown();
+    s_assetManager->Free_Assets();
+    delete s_assetManager;
+    W3D::Shutdown();
+    GameMath::Shutdown();
+    // DX8WebBrowser::Shutdown
+    delete g_theW3DFileSystem;
+}
+
+void Stat_Debug_Display(DebugDisplayInterface *dd, void *v, FILE *f)
+{
+    captainslog_dbgassert(false, "This should never be called directly, but is just a placeholder for drawDebugStats()");
 }
 
 // 0x0073CA80
 void W3DDisplay::Init()
 {
-    // TODO: Requires RTS3DInterfaceScene, RTS2DScene, RTS3DScene, GameAssetManager, ...
-#ifdef GAME_DLL
-    Call_Method<void, W3DDisplay>(0x0073CA80, this);
+    Display::Init();
+
+    if (!m_initialized) {
+        g_theW3DFileSystem = new W3DFileSystem();
+        GameMath::Init();
+
+        s_3DInterfaceScene = new RTS3DInterfaceScene();
+        s_3DInterfaceScene->Set_Ambient_Light(Vector3(1.0f, 1.0f, 1.0f));
+
+        s_2DScene = new RTS2DScene();
+        s_2DScene->Set_Ambient_Light(Vector3(1.0f, 1.0f, 1.0f));
+
+        s_3DScene = new RTS3DScene();
+
+#ifdef GAME_DEBUG_STRUCTS
+        if (g_theWriteableGlobalData->m_wireframe) {
+            s_3DScene->Set_Polygon_Mode(SceneClass::LINE);
+        }
 #endif
+
+        for (int i = 0; i < g_theWriteableGlobalData->m_numberGlobalLights; i++) {
+            m_myLight[i] = new LightClass(LightClass::DIRECTIONAL);
+        }
+
+        Set_Time_Of_Day(g_theWriteableGlobalData->m_timeOfDay);
+
+        for (int i = 0; i < g_theWriteableGlobalData->m_numberGlobalLights; i++) {
+            s_3DScene->Set_Global_Light(m_myLight[i], i);
+        }
+
+        s_assetManager = new GameAssetManager();
+        s_assetManager->Set_W3D_Load_On_Demand(true);
```

Comment:
missing two Register_Prototype_Loader calls

Reply:
Those got moved into the W3DAssetManager constructor.